File: D-MASH/client/backend/node_routing_v4.py
"""Bounded v4 discovery/label forwarding over admitted Node channels.

This runtime owns no Account keys. Store/delegate provisioning and production
endpoint mounting are separate and not yet provided by this module.
"""
import asyncio
import base64
import hashlib
import json
import re
import secrets
import time

if __package__:
    from .probe_primitives_v4 import route_ncrh, extend_ncrh, sample_hop_ttl, consume_hop
    from .recipient_payload_v4 import cover_box
    from .route_discovery_v4 import create_query, answer_query, verify_reply, verify_certificate
else:
    from probe_primitives_v4 import route_ncrh, extend_ncrh, sample_hop_ttl, consume_hop
    from recipient_payload_v4 import cover_box
    from route_discovery_v4 import create_query, answer_query, verify_reply, verify_certificate

TOKEN=re.compile('[0-9a-f]{64}')


def _opaque(value):
    if not isinstance(value,str) or not 96<=len(value)<=21848:raise ValueError('Invalid opaque box')
    raw=base64.b64decode(value,validate=True)
    if not 72<=len(raw)<=16384 or base64.b64encode(raw).decode()!=value:raise ValueError('Invalid opaque box')



class NodeRoutingV4:
    def __init__(self,base_ncrh,*,clock=time.time,monotonic=time.monotonic):
        if not isinstance(base_ncrh,bytes) or len(base_ncrh)!=32:raise ValueError('Invalid BaseNCRH')
        self.base=base_ncrh;self.clock=clock;self.monotonic=monotonic;self.peers={};self.owned=[];self.labels={};self.seen={}
        self.queues={};self.timers={};self.senders={};self.tasks=set();self.pending=set();self.rates={};self.probes=[];self.cover_history=[];self.cover_timer=None;self.cover_policy=None
        self.closed=False;self.stats=dict(forwarded=0,received=0,probes=0,batches=0)

    def _prune(self):
        now=self.clock()
        self.probes[:]=[row for row in self.probes if row[1]['expires_at']>now and row[0] in self.peers]
        for table in (self.labels,self.seen):
            for key,row in list(table.items()):
                if row['expires']<=now:del table[key]

    def _label(self,peer,target,expires):
        self._prune()
        if len(self.labels)>=4096:raise BufferError('Node label quota')
        label=secrets.token_hex(32);self.labels[(peer,label)]={'target':target,'expires':expires}
        return label

    def _dedupe(self,kind,blob,expires):
        self._prune();key=(kind,hashlib.sha256(blob.encode()).digest())
        if key in self.seen:return False
        if len(self.seen)>=4096:raise BufferError('Node duplicate quota')
        self.seen[key]={'expires':expires};return True

    def bind_local(self,certificate,discovery_sign,discovery_box,handler):
        verify_certificate(certificate,int(self.clock()))
        if self.closed or len(self.owned)>=32:raise BufferError('Local route quota')
        if any(row['certificate']['route_id']==certificate['route_id'] for row in self.owned):raise ValueError('Route already bound')
        if discovery_sign.verify_key.encode().hex()!=certificate['discovery_sign'] or bytes(discovery_box.public_key).hex()!=certificate['discovery_box']:
            raise ValueError('Discovery key mismatch')
        binding=dict(certificate=dict(certificate),sign=discovery_sign,box=discovery_box,handler=handler)
        self.owned.append(binding)
        task=asyncio.create_task(self._answer_cached(binding));self.tasks.add(task);task.add_done_callback(self.tasks.discard)
        return binding

    async def _answer_cached(self,binding):
        self._prune()
        for peer,packet in list(self.probes):
            try:await self._answer_binding(peer,packet,binding)
            except (BufferError,ConnectionError):pass

    async def _answer_binding(self,peer,packet,binding):
        if self.closed or peer not in self.peers or packet['expires_at']<=self.clock():return
        try:reply,query_expiry=answer_query(packet['box'],binding['certificate'],binding['sign'],binding['box'],now=int(self.clock()),with_context=True)
        except Exception:return
        expires=min(packet['expires_at'],query_expiry)
        offer=self._label(peer,binding['handler'],expires)
        self._enqueue(peer,dict(type='DATA',version=4,label=packet['return_label'],offer=offer,payload=reply,expires_at=expires))

    def add_peer(self,peer,channel):
        if self.closed or peer in self.peers or len(self.peers)>=8:raise ValueError('Node peer unavailable')
        channel.require_authorized()
        self.peers[peer]=channel
        task=asyncio.create_task(self._read(peer,channel));self.tasks.add(task);task.add_done_callback(self.tasks.discard)
        return task

    async def _read(self,peer,channel):
        try:
            while not self.closed:
                batch=await channel.receive_operation()
                if (not isinstance(batch,dict) or set(batch)!={'type','version','packets'} or batch['type']!='MESH_BATCH'
                        or type(batch['version']) is not int or batch['version']!=4 or not isinstance(batch['packets'],list)
                        or not 1<=len(batch['packets'])<=32 or len(json.dumps(batch))>256*1024):
                    raise ValueError('Invalid Node batch')
                for packet in batch['packets']:
                    channel.require_authorized()
                    await self._receive(peer,packet)
        except (Exception,asyncio.CancelledError):
            pass
        finally:
            if self.peers.get(peer) is channel:self._remove_peer(peer)
            await channel.close()

    def _remove_peer(self,peer):
        self.peers.pop(peer,None);self.queues.pop(peer,None);timer=self.timers.pop(peer,None)
        if timer:timer.cancel()
        self.rates.pop(peer,None)
        if not self.peers:
            for future in self.pending:
                if not future.done():future.set_exception(ConnectionError('Node path unavailable'))
        for key,row in list(self.labels.items()):
            target=row['target']
            if key[0]==peer or (isinstance(target,tuple) and target[0]==peer):del self.labels[key]

    def _enqueue(self,peer,packet):
        if self.closed or peer not in self.peers:raise ConnectionError('Node path unavailable')
        queue=self.queues.setdefault(peer,[])
        size=len(json.dumps(packet,separators=(',',':')))
        if size>32768 or len(queue)>=128 or sum(row[1] for q in self.queues.values() for row in q)+size>1024*1024:
            raise BufferError('Node forwarding queue full')
        queue.append((packet,size,self.monotonic()+0.5))
        if peer not in self.timers and peer not in self.senders:self._arm(peer)

    def _arm(self,peer):
        queue=self.queues.get(peer)
        if not queue or self.closed:return
        delay=max(0,queue[0][2]-self.monotonic())
        self.timers[peer]=asyncio.get_running_loop().call_later(delay,self._start_send,peer)

    def _start_send(self,peer):
        self.timers.pop(peer,None)
        task=asyncio.create_task(self._flush(peer));self.senders[peer]=task

    async def _flush(self,peer):
        try:
            queue=self.queues.get(peer,[])
            # Only the first arrival's window is closed, not later windows.
            deadline=queue[0][2] if queue else 0;batch=[];size=0
            while queue and queue[0][2]<deadline+0.5 and len(batch)<32:
                packet,amount,_=queue[0]
                if size+amount>240*1024:break
                queue.pop(0)
                if packet['expires_at']>self.clock():batch.append(packet);size+=amount
            if batch:
                async with asyncio.timeout(10):
                    await self.peers[peer].send_operation(dict(type='MESH_BATCH',version=4,packets=batch))
                self.stats['batches']+=1
        except Exception:
            channel=self.peers.get(peer);self._remove_peer(peer)
            if channel:await channel.close()
        finally:
            self.senders.pop(peer,None);self._arm(peer)

    async def discover(self,certificate):
        if self.closed or not self.peers:raise ConnectionError('Node path unavailable')
        blob,state=create_query(certificate,now=int(self.clock()));expires=state['query']['expires_at']
        future=asyncio.get_running_loop().create_future();self.pending.add(future)
        async def accept(peer,packet):
            if future.done():return
            try:valid=verify_reply(packet['payload'],state,now=int(self.clock()))
            except Exception:return
            if valid:
                future.set_result(dict(peer=peer,label=packet['offer'],expires_at=min(expires,packet['expires_at']),channel=self.peers.get(peer)))
        try:
            self._dedupe('PROBE',blob,expires)
            ncrh=route_ncrh(self.base,bytes.fromhex(certificate['route_id']));ttl=sample_hop_ttl()
            for peer in list(self.peers):
                label=self._label(peer,accept,expires)
                self._enqueue(peer,dict(type='PROBE',version=4,box=blob,return_label=label,ncrh=ncrh,ttl=ttl,expires_at=expires))
            return await asyncio.wait_for(future,max(0.001,expires-self.clock()))
        finally:
            self.pending.discard(future)
            # Late alternate replies may still arrive on issued capabilities.
            # Retain a no-op target until expiry, without retaining query secrets.
            async def retired(peer,packet):pass
            for row in self.labels.values():
                if row['target'] is accept:row['target']=retired

    def send(self,route,payload,reply_handler):
        expires=route['expires_at']
        if expires<=self.clock() or self.peers.get(route['peer']) is not route['channel']:raise ConnectionError('Node route expired or replaced')
        _opaque(payload)
        offer=self._label(route['peer'],reply_handler,expires)
        self._enqueue(route['peer'],dict(type='DATA',version=4,label=route['label'],offer=offer,payload=payload,expires_at=expires))

    def start_cover(self,*,minimum=15,maximum=45):
        if self.closed or type(minimum) is not int or type(maximum) is not int or not 15<=minimum<maximum<=300:
            raise ValueError('Invalid cover scheduling policy')
        self.stop_cover();self.cover_policy=(minimum,maximum);self._schedule_cover()

    def stop_cover(self):
        self.cover_policy=None
        if self.cover_timer:self.cover_timer.cancel()
        self.cover_timer=None

    def _schedule_cover(self):
        if self.closed or self.cover_policy is None:return
        minimum,maximum=self.cover_policy
        delay=(minimum*1000+secrets.randbelow((maximum-minimum)*1000+1))/1000
        self.cover_timer=asyncio.get_running_loop().call_later(delay,self._cover_tick)

    def _cover_tick(self):
        self.cover_timer=None
        if self.closed or self.cover_policy is None:return
        try:self.inject_cover_once()
        except (ValueError,ConnectionError,BufferError):pass
        finally:self._schedule_cover()

    def inject_cover_once(self,size=1024):
        if self.closed or any(self.queues.values()) or self.senders:return False
        now=self.monotonic();self.cover_history[:]=[row for row in self.cover_history if row[0]>now-60]
        if type(size) is not int or not 256<=size<=16384:raise ValueError('Invalid cover size')
        if len(self.cover_history)>=4 or sum(row[2] for row in self.cover_history)+size>16384:return False
        self._prune();candidates=[]
        for row in self.labels.values():
            target=row['target']
            if (isinstance(target,tuple) and target[0] in self.peers and row['expires']>self.clock()+2
                    and sum(row[1]==target[0] for row in self.cover_history)<2):
                candidates.append((target,row['expires']))
        if not candidates:return False
        target,expires=secrets.choice(candidates)
        async def discard(peer,packet):pass
        self.send(dict(peer=target[0],label=target[1],expires_at=expires,channel=self.peers[target[0]]),cover_box(size),discard)
        self.cover_history.append((now,target[0],size))
        return True

    async def _receive(self,peer,packet):
        if not isinstance(packet,dict) or packet.get('version')!=4 or type(packet.get('version')) is not int:raise ValueError('Invalid Node packet')
        expires=packet.get('expires_at')
        if type(expires) is not int or expires>self.clock()+180:raise ValueError('Invalid packet expiry')
        if expires<=self.clock():return
        if packet.get('type')=='PROBE':
            if set(packet)!={'type','version','box','return_label','ncrh','ttl','expires_at'}:raise ValueError('Invalid Probe')
            for key in ('return_label','ncrh'):
                if not isinstance(packet[key],str) or not TOKEN.fullmatch(packet[key]):raise ValueError('Invalid Probe label')
            remaining=consume_hop(packet['ttl']);blob=packet['box']
            _opaque(blob)
            bucket=int(self.clock()//60);rate=self.rates.get(peer,(bucket,0));count=rate[1] if rate[0]==bucket else 0
            if count>=32:raise BufferError('Probe rate limit')
            self.rates[peer]=(bucket,count+1)
            if not self._dedupe('PROBE',blob,expires):return
            self.stats['probes']+=1
            self.probes.append((peer,dict(packet)));self.probes[:]=self.probes[-64:]
            if remaining:
                ncrh=extend_ncrh(self.base,packet['ncrh'])
                for neighbor in list(self.peers):
                    if neighbor==peer:continue
                    label=self._label(neighbor,(peer,packet['return_label']),expires)
                    self._enqueue(neighbor,{**packet,'return_label':label,'ttl':remaining,'ncrh':ncrh})
            for binding in self.owned:await self._answer_binding(peer,packet,binding)
        elif packet.get('type')=='DATA':
            if set(packet)!={'type','version','label','offer','payload','expires_at'}:raise ValueError('Invalid DATA')
            for key in ('label','offer'):
                if not isinstance(packet[key],str) or not TOKEN.fullmatch(packet[key]):raise ValueError('Invalid DATA label')
            _opaque(packet['payload'])
            self._prune();binding=self.labels.get((peer,packet['label']))
            if not binding or expires>binding['expires']:return
            if not self._dedupe('DATA',packet['payload'],expires):return
            target=binding['target'];self.stats['received']+=1
            if isinstance(target,tuple):
                offer=self._label(target[0],(peer,packet['offer']),expires)
                self._enqueue(target[0],{**packet,'label':target[1],'offer':offer})
                self.stats['forwarded']+=1
            else:
                await target(peer,packet)
        else:raise ValueError('Unsupported Node packet')

    async def close(self):
        self.closed=True;self.stop_cover()
        for future in self.pending:
            if not future.done():future.set_exception(ConnectionError('Node runtime closed'))
        for timer in self.timers.values():timer.cancel()
        tasks=list(self.tasks)+list(self.senders.values())
        for task in tasks:task.cancel()
        await asyncio.gather(*tasks,return_exceptions=True)
        self.peers.clear();self.labels.clear();self.queues.clear();self.seen.clear();self.owned.clear();self.rates.clear();self.probes.clear()

File: D-MASH PWA/not_messenger/js/node_routing_v4.js
'use strict';
(function(global){
 const token=value=>typeof value==='string'&&/^[0-9a-f]{64}$/.test(value);
 const exact=(value,keys)=>value&&typeof value==='object'&&!Array.isArray(value)&&Object.keys(value).sort().join(',')===keys;
 const hex=value=>Array.from(value,b=>b.toString(16).padStart(2,'0')).join('');
 const opaque=value=>{
  if(typeof value!=='string'||value.length<96||value.length>21848)throw Error('Invalid opaque box');
  const raw=atob(value);if(raw.length<72||raw.length>16384||btoa(raw)!==value)throw Error('Invalid opaque box');
 };
 const labelKey=(peer,label)=>JSON.stringify([peer,label]);
 const randomLabel=()=>hex(crypto.getRandomValues(new Uint8Array(32)));
 class NodeRoutingV4{
  constructor(base,{clock=()=>Date.now()/1000,monotonic=()=>performance.now()/1000}={}){
   if(!(base instanceof Uint8Array)||base.length!==32)throw Error('Invalid BaseNCRH');
   this.base=base.slice();this.clock=clock;this.monotonic=monotonic;this.peers=new Map();this.owned=[];this.labels=new Map();this.seen=new Map();
   this.queues=new Map();this.timers=new Map();this.senders=new Map();this.tasks=new Set();this.pending=new Set();this.rates=new Map();this.probes=[];
   this.closed=false;this.stats={forwarded:0,received:0,probes:0,batches:0};
  }
  prune(){
   const now=this.clock();this.probes=this.probes.filter(row=>row.packet.expires_at>now&&this.peers.has(row.peer));for(const table of [this.labels,this.seen])for(const [key,row] of table)if(row.expires<=now)table.delete(key);
  }
  label(peer,target,expires){
   this.prune();if(this.labels.size>=4096)throw Error('Node label quota');
   const label=randomLabel();this.labels.set(labelKey(peer,label),{peer,target,expires});return label;
  }
  async dedupe(kind,blob,expires){
   const hash=hex(new Uint8Array(await crypto.subtle.digest('SHA-256',new TextEncoder().encode(blob))));
   if(this.closed)throw Error('Node runtime closed');
   this.prune();const key=kind+hash;if(this.seen.has(key))return false;
   if(this.seen.size>=4096)throw Error('Node duplicate quota');this.seen.set(key,{expires});return true;
  }
  bindLocal(certificate,sign,box,handler){
   global.DmashRouteDiscoveryV4.verifyCertificate(certificate,Math.floor(this.clock()));
   if(this.closed||this.owned.length>=32)throw Error('Local route unavailable');
   if(hex(sign.publicKey)!==certificate.discovery_sign||hex(box.publicKey)!==certificate.discovery_box)throw Error('Discovery key mismatch');
   if(this.owned.some(row=>row.certificate.route_id===certificate.route_id))throw Error('Route already bound');
   const binding={certificate:{...certificate},sign,box,handler};this.owned.push(binding);
   const task=this.answerCached(binding);this.tasks.add(task);task.finally(()=>this.tasks.delete(task));return binding;
  }
  async answerCached(binding){
   this.prune();for(const row of [...this.probes])try{await this.answerBinding(row.peer,row.packet,binding);}catch(_){}
  }
  async answerBinding(peer,packet,binding){
   if(this.closed||!this.peers.has(peer)||packet.expires_at<=this.clock())return;
   let answer;try{answer=await global.DmashRouteDiscoveryV4.answerQuery(packet.box,binding.certificate,binding.sign,binding.box,{clock:()=>Math.floor(this.clock()),withContext:true});}catch(_){return;}
   if(this.closed||!this.peers.has(peer)||packet.expires_at<=this.clock())return;
   const expires=Math.min(packet.expires_at,answer.expiresAt),offer=this.label(peer,binding.handler,expires);
   this.enqueue(peer,{type:'DATA',version:4,label:packet.return_label,offer,payload:answer.reply,expires_at:expires});
  }
  async addPeer(peer,channel){
   if(this.closed||this.peers.has(peer)||this.peers.size>=8)throw Error('Node peer unavailable');
   await channel.requireAuthorized();
   if(this.closed||this.peers.has(peer)||this.peers.size>=8)throw Error('Node peer unavailable');
   this.peers.set(peer,channel);const task=this.read(peer,channel);this.tasks.add(task);task.finally(()=>this.tasks.delete(task));
  }
  async read(peer,channel){
   try{
    while(!this.closed){
     const batch=await channel.receiveOperation();
     if(!exact(batch,'packets,type,version')||batch.type!=='MESH_BATCH'||batch.version!==4||!Array.isArray(batch.packets)||batch.packets.length<1||batch.packets.length>32||JSON.stringify(batch).length>256*1024)throw Error('Invalid Node batch');
     for(const packet of batch.packets){await channel.requireAuthorized();await this.receive(peer,packet);}
    }
   }catch(_){}
   finally{if(this.peers.get(peer)===channel)this.removePeer(peer);channel.close();}
  }
  removePeer(peer){
   this.peers.delete(peer);this.queues.delete(peer);clearTimeout(this.timers.get(peer));this.timers.delete(peer);this.rates.delete(peer);
   for(const [key,row] of this.labels)if(row.peer===peer||(Array.isArray(row.target)&&row.target[0]===peer))this.labels.delete(key);
   if(!this.peers.size)for(const pending of this.pending)pending.reject(Error('Node path unavailable'));
  }
  enqueue(peer,packet){
   if(this.closed||!this.peers.has(peer))throw Error('Node path unavailable');
   const queue=this.queues.get(peer)||[];this.queues.set(peer,queue);
   const size=JSON.stringify(packet).length;
   let total=0;for(const q of this.queues.values())for(const row of q)total+=row.size;
   if(size>32768||queue.length>=128||total+size>1024*1024)throw Error('Node forwarding queue full');
   queue.push({packet,size,deadline:this.monotonic()+0.5});
   if(!this.timers.has(peer)&&!this.senders.has(peer))this.arm(peer);
  }
  arm(peer){
   const queue=this.queues.get(peer);if(!queue?.length||this.closed)return;
   this.timers.set(peer,setTimeout(()=>this.startSend(peer),Math.max(0,queue[0].deadline-this.monotonic())*1000));
  }
  startSend(peer){this.timers.delete(peer);const task=Promise.resolve().then(()=>this.flush(peer));this.senders.set(peer,task);}
  async flush(peer){
   let timer;
   try{
    const queue=this.queues.get(peer)||[],deadline=queue[0]?.deadline||0,batch=[];let size=0;
    while(queue.length&&queue[0].deadline<deadline+0.5&&batch.length<32){
     const row=queue[0];if(size+row.size>240*1024)break;queue.shift();
     if(row.packet.expires_at>this.clock()){batch.push(row.packet);size+=row.size;}
    }
    if(batch.length){
     await Promise.race([this.peers.get(peer).sendOperation({type:'MESH_BATCH',version:4,packets:batch}),new Promise((_,reject)=>{timer=setTimeout(()=>reject(Error('Node send timeout')),10000);})]);
     this.stats.batches++;
    }
   }catch(_){const channel=this.peers.get(peer);this.removePeer(peer);channel?.close();}
   finally{clearTimeout(timer);this.senders.delete(peer);this.arm(peer);}
  }
  async discover(certificate){
   if(this.closed||!this.peers.size)throw Error('Node path unavailable');
   const {blob,state}=global.DmashRouteDiscoveryV4.createQuery(certificate,{now:Math.floor(this.clock())}),expires=state.query.expires_at;
   let resolve,reject,timer,done=false;
   const result=new Promise((yes,no)=>{resolve=value=>{done=true;yes(value);};reject=error=>{done=true;no(error);};});
   // An early disconnect may reject while asynchronous NCRH work is pending.
   result.catch(()=>{});const pending={reject};this.pending.add(pending);
   const accept=async(peer,packet)=>{
    if(done)return;
    const valid=await global.DmashRouteDiscoveryV4.verifyReply(packet.payload,state,{clock:()=>Math.floor(this.clock())});
    if(valid&&!done&&!this.closed&&expires>this.clock()&&this.peers.has(peer))resolve({peer,label:packet.offer,expires_at:Math.min(expires,packet.expires_at),channel:this.peers.get(peer)});
   };
   try{
    await this.dedupe('PROBE',blob,expires);
    const ncrh=await global.DmashProbePrimitivesV4.routeNcrh(this.base,Uint8Array.from(certificate.route_id.match(/../g),b=>parseInt(b,16)));
    const ttl=global.DmashProbePrimitivesV4.sampleHopTtl();
    for(const peer of this.peers.keys()){
     const label=this.label(peer,accept,expires);this.enqueue(peer,{type:'PROBE',version:4,box:blob,return_label:label,ncrh,ttl,expires_at:expires});
    }
    timer=setTimeout(()=>reject(Error('Node discovery expired')),Math.max(1,(expires-this.clock())*1000));
    return await result;
   }finally{
    done=true;clearTimeout(timer);this.pending.delete(pending);state.replyPrivate.fill(0);
    for(const row of this.labels.values())if(row.target===accept)row.target=async()=>{};
   }
  }
  send(route,payload,replyHandler){
   if(route.expires_at<=this.clock()||this.peers.get(route.peer)!==route.channel)throw Error('Node route expired or replaced');
   opaque(payload);const offer=this.label(route.peer,replyHandler,route.expires_at);
   this.enqueue(route.peer,{type:'DATA',version:4,label:route.label,offer,payload,expires_at:route.expires_at});
  }
  async receive(peer,packet){
   if(!packet||packet.version!==4||!Number.isSafeInteger(packet.expires_at)||packet.expires_at>this.clock()+180)throw Error('Invalid Node packet');
   const expires=packet.expires_at;if(expires<=this.clock())return;
   if(packet.type==='PROBE'){
    if(!exact(packet,'box,expires_at,ncrh,return_label,ttl,type,version')||!token(packet.return_label)||!token(packet.ncrh))throw Error('Invalid Probe');
    const remaining=global.DmashProbePrimitivesV4.consumeHop(packet.ttl),blob=packet.box;
    opaque(blob);
    const bucket=Math.floor(this.clock()/60),rate=this.rates.get(peer),count=rate?.bucket===bucket?rate.count:0;
    if(count>=32)throw Error('Probe rate limit');this.rates.set(peer,{bucket,count:count+1});
    if(!await this.dedupe('PROBE',blob,expires)||!this.peers.has(peer))return;this.stats.probes++;
    this.probes.push({peer,packet:{...packet}});this.probes=this.probes.slice(-64);
    if(remaining){
     const ncrh=await global.DmashProbePrimitivesV4.extendNcrh(this.base,packet.ncrh);
     if(this.closed||!this.peers.has(peer))return;
     for(const neighbor of this.peers.keys())if(neighbor!==peer){
      const label=this.label(neighbor,[peer,packet.return_label],expires);this.enqueue(neighbor,{...packet,return_label:label,ttl:remaining,ncrh});
     }
    }
    for(const binding of this.owned)await this.answerBinding(peer,packet,binding);
   }else if(packet.type==='DATA'){
    if(!exact(packet,'expires_at,label,offer,payload,type,version')||!token(packet.label)||!token(packet.offer))throw Error('Invalid DATA');
    opaque(packet.payload);this.prune();const binding=this.labels.get(labelKey(peer,packet.label));if(!binding||expires>binding.expires)return;
    if(!await this.dedupe('DATA',packet.payload,expires)||this.labels.get(labelKey(peer,packet.label))!==binding||!this.peers.has(peer))return;this.stats.received++;
    const target=binding.target;
    if(Array.isArray(target)){
     const offer=this.label(target[0],[peer,packet.offer],expires);this.enqueue(target[0],{...packet,label:target[1],offer});this.stats.forwarded++;
    }else await target(peer,packet);
   }else throw Error('Unsupported Node packet');
  }
  async close(){
   this.closed=true;
   for(const pending of this.pending)pending.reject(Error('Node runtime closed'));
   for(const timer of this.timers.values())clearTimeout(timer);
   for(const channel of this.peers.values())channel.close();
   await Promise.allSettled([...this.tasks,...this.senders.values()]);
   this.peers.clear();this.labels.clear();this.queues.clear();this.seen.clear();this.owned=[];this.rates.clear();this.probes=[];this.base.fill(0);
  }
 }
 global.DmashNodeRoutingV4=NodeRoutingV4;
 if(typeof module!=='undefined')module.exports=NodeRoutingV4;
})(typeof window!=='undefined'?window:globalThis);
